### PR TITLE
Optimize Enumerable.OfType with custom iterator

### PR DESCRIPTION
**Pull request by stephentoub**
| Method              | Toolchain         | Target | Mean        | Ratio | Allocated | Alloc Ratio |
|-------------------- |------------------ |------- |------------:|------:|----------:|------------:|
| OfTypeForEach       | \main\corerun.exe | Object |   112.64 ns |  1.00 |      88 B |        1.00 |
| OfTypeForEach       | \pr\corerun.exe   | Object |    66.18 ns |  0.59 |      80 B |        0.91 |
|                     |                   |        |             |       |           |             |
| OfTypeCount         | \main\corerun.exe | Object |   119.40 ns |  1.00 |      88 B |        1.00 |
| OfTypeCount         | \pr\corerun.exe   | Object |   101.25 ns |  0.85 |      80 B |        0.91 |
|                     |                   |        |             |       |           |             |
| OfTypeToArray       | \main\corerun.exe | Object |   207.82 ns |  1.00 |     160 B |        1.00 |
| OfTypeToArray       | \pr\corerun.exe   | Object |   129.60 ns |  0.62 |     152 B |        0.95 |
|                     |                   |        |             |       |           |             |
| OfTypeToList        | \main\corerun.exe | Object |   214.81 ns |  1.00 |     264 B |        1.00 |
| OfTypeToList        | \pr\corerun.exe   | Object |   158.42 ns |  0.74 |     256 B |        0.97 |
|                     |                   |        |             |       |           |             |
| OfTypeFirst         | \main\corerun.exe | Object |    46.60 ns |  1.00 |      88 B |        1.00 |
| OfTypeFirst         | \pr\corerun.exe   | Object |    47.11 ns |  1.01 |      80 B |        0.91 |
|                     |                   |        |             |       |           |             |
| OfTypeLast          | \main\corerun.exe | Object |   146.69 ns |  1.00 |      88 B |        1.00 |
| OfTypeLast          | \pr\corerun.exe   | Object |    72.18 ns |  0.49 |      80 B |        0.91 |
|                     |                   |        |             |       |           |             |
| OfTypeSelectForEach | \main\corerun.exe | Object |   172.42 ns |  1.00 |     144 B |        1.00 |
| OfTypeSelectForEach | \pr\corerun.exe   | Object |    84.93 ns |  0.49 |     104 B |        0.72 |
|                     |                   |        |             |       |           |             |
| OfTypeForEach       | \main\corerun.exe | String | 2,175.93 ns |  1.00 |      88 B |        1.00 |
| OfTypeForEach       | \pr\corerun.exe   | String | 1,046.87 ns |  0.48 |      80 B |        0.91 |
|                     |                   |        |             |       |           |             |
| OfTypeCount         | \main\corerun.exe | String | 2,065.02 ns |  1.00 |      88 B |        1.00 |
| OfTypeCount         | \pr\corerun.exe   | String |   974.42 ns |  0.47 |      80 B |        0.91 |
|                     |                   |        |             |       |           |             |
| OfTypeToArray       | \main\corerun.exe | String | 3,289.18 ns |  1.00 |    1432 B |        1.00 |
| OfTypeToArray       | \pr\corerun.exe   | String | 4,942.53 ns |  1.50 |    1424 B |        0.99 |
|                     |                   |        |             |       |           |             |
| OfTypeToList        | \main\corerun.exe | String | 4,016.68 ns |  1.00 |    4352 B |        1.00 |
| OfTypeToList        | \pr\corerun.exe   | String | 2,926.28 ns |  0.73 |    4344 B |        1.00 |
|                     |                   |        |             |       |           |             |
| OfTypeFirst         | \main\corerun.exe | String |    44.71 ns |  1.00 |      88 B |        1.00 |
| OfTypeFirst         | \pr\corerun.exe   | String |    31.20 ns |  0.70 |      80 B |        0.91 |
|                     |                   |        |             |       |           |             |
| OfTypeLast          | \main\corerun.exe | String | 2,226.84 ns |  1.00 |      88 B |        1.00 |
| OfTypeLast          | \pr\corerun.exe   | String | 1,339.43 ns |  0.59 |      80 B |        0.91 |
|                     |                   |        |             |       |           |             |
| OfTypeSelectForEach | \main\corerun.exe | String | 2,689.61 ns |  1.00 |     144 B |        1.00 |
| OfTypeSelectForEach | \pr\corerun.exe   | String | 1,150.00 ns |  0.43 |     104 B |        0.72 |

```C#
using BenchmarkDotNet.Attributes;
using BenchmarkDotNet.Running;
using System.Reflection;

BenchmarkSwitcher.FromAssembly(typeof(Tests).Assembly).Run(args);

[MemoryDiagnoser(false)]
[HideColumns("Job", "Error", "StdDev", "Median", "RatioSD")]
public partial class Tests
{
    private IEnumerable<MemberInfo> _members;

    [Params(typeof(object), typeof(string))]
    public Type Target { get; set; }

    [GlobalSetup]
    public void Setup() => _members = Target.GetMembers();

    [Benchmark]
    public int OfTypeForEach()
    {
        int sum = 0;
        foreach (MethodInfo member in _members.OfType<MethodInfo>())
        {
            sum += member.Name.Length;
        }
        return sum;
    }

    [Benchmark]
    public int OfTypeCount() => _members.OfType<MethodInfo>().Count();

    [Benchmark]
    public MethodInfo[] OfTypeToArray() => _members.OfType<MethodInfo>().ToArray();

    [Benchmark]
    public List<MethodInfo> OfTypeToList() => _members.OfType<MethodInfo>().ToList();

    [Benchmark]
    public MethodInfo OfTypeFirst() => _members.OfType<MethodInfo>().First();

    [Benchmark]
    public MethodInfo OfTypeLast() => _members.OfType<MethodInfo>().Last();

    [Benchmark]
    public int OfTypeSelectForEach()
    {
        int sum = 0;
        foreach (string name in _members.OfType<MethodInfo>().Select(m => m.Name))
        {
            sum += name.Length;
        }
        return sum;
    }
}
```

### CODE REVIEW

File: src/libraries/System.Linq/src/System/Linq/Cast.SpeedOpt.cs [hunk 1]
```diff
@@ -3,11 +3,175 @@
...
+#else
+                        Unsafe.As<Func<object, TResult2>>(selector);
+#endif
+
+                    // We can special-case arrays and IEnumerable to use the corresponding WhereSelectIterators because
+                    // they're covariant. It's not worthwhile checking for List<T> to use the ListWhereSelectIterator
+                    // because List<> is not covariant.
+                    Func<object, bool> isTResult = static o => o is TResult;
+                    return objectSource is object[] array ?
+                        new ArrayWhereSelectIterator<object, TResult2>(array, isTResult, localSelector) :
```
Comment (eiriktsarpalis) [inline src/libraries/System.Linq/src/System/Linq/Cast.SpeedOpt.cs:167]: Could allocating the intermediate iterator be avoid somehow? E.g. by making the particular `Select` implementations static if possible?

File: src/libraries/System.Linq/src/System/Linq/Cast.SpeedOpt.cs [hunk 2]
```diff
@@ -3,11 +3,175 @@
...
 using System.Collections;
 using System.Collections.Generic;
+using System.Diagnostics;
+using System.Runtime.CompilerServices;
 
 namespace System.Linq
 {
     public static partial class Enumerable
     {
+        private sealed partial class OfTypeIterator<TResult>
```
Comment (eiriktsarpalis) [inline src/libraries/System.Linq/src/System/Linq/Cast.SpeedOpt.cs:13]: Nit: does the added logic justify extracting to separate `OfType*.cs` files?

File: src/libraries/System.Linq.Parallel/tests/QueryOperators/AsEnumerableTests.cs [hunk 3]
```diff
@@ -50,12 +50,13 @@ public static void AsEnumerable_LinqBinding(Labeled<ParallelQuery<int>> labeled,
...
             IEnumerable<int> enumerable = labeled.Item.AsEnumerable();
 
-            // The LINQ Cast<T>() retains origin type for ParallelEnumerable  and Partitioner when unordered,
+            // The LINQ Cast<T>() retains origin type for ParallelEnumerable and Partitioner when unordered,
             // (and all when ordered, due to the extra wrapper)
             // although aliased as IEnumerable<T>, so further LINQ calls work as expected.
             // If this test starts failing, update this test, and maybe mention it in release notes.
             Assert.IsNotType<ParallelQuery<int>>(enumerable.Cast<int>());
             Assert.True(enumerable.Cast<int>() is ParallelQuery<int>);
+            Assert.True(enumerable.OfType<int>() is ParallelQuery<int>); // for non-nullable value types, OfType is equivalent to Cast
```
Comment (eiriktsarpalis) [inline src/libraries/System.Linq.Parallel/tests/QueryOperators/AsEnumerableTests.cs:59]: Presumably PLINQ can't access `Iterator<T>` types, so what specifically about this change resulted in the change in behavior?

**Review by eiriktsarpalis — APPROVED** [review]

File: src/libraries/System.Linq.Parallel/tests/QueryOperators/AsEnumerableTests.cs (hunk 3 above)
Comment (stephentoub) [reply src/libraries/System.Linq.Parallel/tests/QueryOperators/AsEnumerableTests.cs:59]: > so what specifically about this change resulted in the change in behavior?

This:
https://github.com/dotnet/runtime/pull/99216/files#diff-9d518d344f6408b49ba2852e96d33eb701d71b93e5c83444731fbd3fd4c7ab25R19-R24

File: src/libraries/System.Linq/src/System/Linq/Cast.SpeedOpt.cs (hunk 1 above)
Comment (stephentoub) [reply src/libraries/System.Linq/src/System/Linq/Cast.SpeedOpt.cs:167]: Which intermediate are you referring to? At this point we already have the OfType iterator, and now that someone is calling Select, we have to return a new object; mutating the current object would be visible and dangerous.

File: src/libraries/System.Linq/src/System/Linq/Cast.SpeedOpt.cs (hunk 2 above)
Comment (stephentoub) [reply src/libraries/System.Linq/src/System/Linq/Cast.SpeedOpt.cs:13]: If you'd prefer that, I can do so, or we can do it subsequently.

Comment (eiriktsarpalis) [reply src/libraries/System.Linq/src/System/Linq/Cast.SpeedOpt.cs:13]: I think keeping it separate would be nice to have. The methods are similar but have distinct semantics and AFAICT they don't share logic.

File: src/libraries/System.Linq/src/System/Linq/Cast.SpeedOpt.cs (hunk 1 above)
Comment (eiriktsarpalis) [reply src/libraries/System.Linq/src/System/Linq/Cast.SpeedOpt.cs:167]: I misunderstood what this method is doing, nevermind.

**Pull request merged**